src/librustc_mir/hair/pattern/mod.rs

```diff
@@ -1096,7 +1101,11 @@ pub fn compare_const_vals(a: &ConstVal, b: &ConstVal, ty: Ty) -> Option<Ordering
                     // FIXME(oli-obk): report cmp errors?
                     l.try_cmp(r).ok()
                 },
-                ty::TyInt(_) => Some((a as i128).cmp(&(b as i128))),
+                ty::TyInt(_) => {
+                    let a = interpret::sign_extend(tcx, a, ty).expect("layout error for TyInt");
```

Comment:
this is the only actual change in this PR. Everything else is just moving things around.

The issue was that we represent constants in their native bit width, so casting to i128 didn't work out for e.g. `-1u8`, which is `0xFF`, and stays that when casting to `i128` (meaning it would be 255 instead of -1)